docs/proposals/blocks-storage-bucket-index.md

```diff
@@ -0,0 +1,109 @@
+---
+title: "Blocks storage bucket index"
+linkTitle: "Blocks storage bucket index"
+weight: 1
+slug: blocks-storage-bucket-index
+---
+
+- Author: [Marco Pracucci](https://github.com/pracucci)
+- Date: November 2020
+- Status: draft
+
+## Introduction
+
+Queriers and store-gateways, at startup and periodically while running, need to scan the entire object store bucket to find tenants and blocks for each tenants.
+
+For each discovered block, they need to fetch the `meta.json` file and check for the existence of `deletion-mark.json` (used to signal a block is "marked for deletion"). The `meta.json` file is immutable and gets cached, but `deletion-mark.json` can be created at any time and its non-existence can't be safely cached for long (actually our cache is pretty ineffective regarding this).
+
+The number of blocks in the storage linearly increases with the number of tenants, and so does the number of bucket API calls required to scan the bucket.
+
+For example, assuming 1 block / day / tenant and a 400d retention, we would have 400 blocks for a single-tenant cluster, but 4M blocks for a 10K tenants cluster (regardless of the number of active series or their QPS).
+
+We're currently testing a cluster running with 6K tenants and we have faced the following issues:
+
+- The cold start of queriers and store-gateways (empty caches) take tens of minutes.
+- If we increase the concurrency of bucket operations run to scan the bucket, we hit into object store rate limits and/or the 5xx error rate increases
+- The need to run "list objects" API call for queriers and store-gateways to discover blocks and the check for the existence of `deletion-mark.json` (which requires a read operation for each block and can't be cached for a long time) represents a significant % of bucket API calls baseline costs, regardless the tenants QPS (costs we have even if the cluster has 0 queries)
+
+## Goal
+
+The goal of this proposal is:
+
+- The querier should be up and running without having to scan the bucket at all (zero startup time)
+- The querier should not run any "list objects" operation at anytime
+- The querier should require only 1 "get object" operation to get the entire view of a tenant's blocks
+
+## Out of scope
+
+We believe the same technique described in this proposal could be applied to optimise the store-gateway too (we've built a [PoC](https://github.com/cortexproject/cortex/compare/pracucci:experiment-object-store-based-blocks-index)), however to keep the design doc easier to discuss we suggest to keep the store-gateway out of scope and address it in a follow-up design doc.
```

Comment:
Is there any reason the store-gateway would need significant changes to how the index is loaded and used?

Reply:
Not significant changes, but the problem is that the store-gateway would have to preload all index-headers anyway, so the main objection would be "preloading all index-headers is still a pain, even with the bucket index in place". I have an idea to not preload all index-headers in the store-gateway, but I would prefer to discuss it in a separate design doc, to keep this doc lighter.